### PR TITLE
Math packages: gf2x, ntl, flintlib, arb, eclib

### PR DESCRIPTION
**Pull request by tornaria**
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me

#### Notes
- This includes: `gf2x` (see also #29783), `ntl`, `flintlib`, `arb`, `eclib`.
- Everything is tested and check pass in x86_64, x86_64-musl, i686.

EDIT: split in smaller independent PR to ease review.

### CODE REVIEW

**Comment by abenson** [comment]
> This continues #29997.

Just add to that one.  At this point the two conflict -- it's easier to track when it's all in the same PR.

**Comment by tornaria** [comment]
> > This continues #29997.
> 
> Just add to that one. At this point the two conflict -- it's easier to track when it's all in the same PR.

You are right. I rebased in several different and independent PR.

**Comment by dkwo** [comment]
I did not test `eclib`, the rest looks good to me.

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:5]:
```suggestion
build_style=configure
```

Style nit

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:14]: Do you have  a simple comment you can add here explaining why?

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:18]: Doesn't need to be conditional, setting this for glibc shouldn't hurt anything.

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:22]:
```suggestion
	depends+=" ${makedepends}"
```

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:21]:
```suggestion
	depends="arb>=${version}_${revision}"
```

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:25]:
```suggestion
		vmove usr/include
```

Style nit

File: srcpkgs/eclib/template
Comment (ericonr) [inline srcpkgs/eclib/template:17]: We usually try to use `autoreconf -fi`, does their script do more than that?

File: srcpkgs/eclib/template
Comment (ericonr) [inline srcpkgs/eclib/template:21]:
```suggestion
	depends="eclib>=${version}_${revision}"
```

File: srcpkgs/eclib/template
Comment (ericonr) [inline srcpkgs/eclib/template:28]: Same style nit, remove quotes when not globbing.

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:19]: Same as previous

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:15]: Same as previous

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:26]: Same as previous

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:22]:
```suggestion
	depends="flintlib>=${version}_${revision}"
```

File: srcpkgs/flintlib/update
Comment (ericonr) [inline srcpkgs/flintlib/update:1]: Maybe mention in commit that name was picked because of existing package named flint.

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:6]:
```suggestion
build_style=gnu-configure
```

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:17]:
```suggestion
if [ -z "$build_option_native" ]; then
```

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:15]: Please add a description for this option, it isn't obvious what it does. `desc_option_native="..."`

I would name it sse3 directly, maybe. Then you could do `configure_args="$(vopt_enable sse3)"`.

Even if you keep the native name, you can still do `configure_args="$(vopt_if native '' '--disable-sse3')"` and have configure args together with build style instead of down here.

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:26]:
```suggestion
	depends="gf2x>=${version}_${revision}"
```

Comment (ericonr) [reply srcpkgs/gf2x/template:26]: I remembered it just now, but the default style is actually `${sourcepkg}>=...`

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:32]: Same as above

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:6]: Same as above

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:15]: Same as above

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:17]: Same as above

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:27]: Same as above

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:34]: Same as above.

File: srcpkgs/ntl/update
Comment (ericonr) [inline srcpkgs/ntl/update:1]: Maybe comment in template that homepage is used in update file too? So it isn't clobbered accidentally.

**Review by ericonr — COMMENTED** [review]
Remember to squash all the commits for eclib :)

File: srcpkgs/arb/template
Comment (tornaria) [reply srcpkgs/arb/template:21]: Somehow I feel uneasy about that.

Comment (ericonr) [reply srcpkgs/arb/template:21]: This is done because otherwise parsing is ambiguous and that sucks.

Comment (ericonr) [reply srcpkgs/arb/template:21]: And it helps with updates, in some occasions

File: srcpkgs/gf2x/template
Comment (tornaria) [reply srcpkgs/gf2x/template:15]: I added this to `common/options.description` in #29997 (see 2274429). There is a discussion about this in that PR.

My intention is to have an easy way to recompile math software for native arch. Depending on the package sometimes is sse3, sometimes is more. In this case the library also uses sse4.1 and pclmul, but they are dependents so `--disable-sse3` is enough to disable all of that is not generic (x86_64 implies sse2 but nothing more).

Not disabling sse3 as above means that the library will be compiled with -msse3, -msse41, -mpclmul, ... and maybe also -march=native, etc, so that all the options available in the current cpu will be used.

In general, different software have different features, oftentimes they have custom written asm for different architectures so selecting build option "native" will not necessarily be equivalent to just using `-march=native`.

In many cases (like here) one has to be careful because native is the default and it is necessary to make sure the binary actually is generic enough for x86_64 (I'm not sure how to test this -- can a VM disable some sse levels even if the host has them? I have an old core 2 quad (q9550) working (even with 8GB RAM) but that already has sse3!

Comment (ericonr) [reply srcpkgs/gf2x/template:15]: AFAIK you can test, but I'm not sure it's necessary. We can wait for bug reports, IMO.

 Please import the commit into this PR, then :)

Comment (ericonr) [reply srcpkgs/gf2x/template:15]: I wonder if `native_build` would be better.

**Comment by tornaria** [comment]
General comment for @ericonr : Thanks for your detailed review. I will apply all suggestions you made, and update to newer versions for arb, eclib, flintlib, ntl.

I hope this can be merged into void -- I want to help include more math software in void, I can maintain, update, etc.

File: srcpkgs/arb/template
Comment (tornaria) [reply srcpkgs/arb/template:14]: I don't remember from the top of my head; I did try it and something didn't work. I will retry it and comment.

File: srcpkgs/arb/template
Comment (tornaria) [reply srcpkgs/arb/template:22]: Note that many times `makedepends` has more packages than `depends` for the -devel package, just not in this case, that's why I tried to be explicit about it.

File: srcpkgs/eclib/template
Comment (tornaria) [reply srcpkgs/eclib/template:17]:
```
aclocal
libtoolize --automake
automake --add-missing
autoconf
```

I'll try to check if `autoreconf -fi` is enough.

File: srcpkgs/ntl/update
Comment (tornaria) [reply srcpkgs/ntl/update:1]: Isn't `"${homepage}"` the default for `site` in any case?

Comment (ericonr) [reply srcpkgs/ntl/update:1]: Right, but if homepage moves to a github website or whatever the path won't exist anymore.

File: srcpkgs/arb/template
Comment (ericonr) [reply srcpkgs/arb/template:22]: Ok, that's fair. Can leave as is.

**Comment by ericonr** [comment]
My `_GNU_SOURCE` suggestion was apparently a bad idea for one of the packages, feel free to move it back to a conditional for musl.

Turns out glibc will define its own `fdiv` function if you include `<tgmath.h>` with `_GNU_SOURCE` defined, so we can't define it for glibc to avoid polluting the namespace and causing conflicts.

**Comment by tornaria** [comment]
> My `_GNU_SOURCE` suggestion was apparently a bad idea for one of the packages, feel free to move it back to a conditional for musl.

I actually went ahead and fixed it at the source: https://github.com/wbhart/flint2/pull/988 and https://github.com/fredrik-johansson/arb/pull/376. My new version will include those as patches which they will hopefully be merged upstream.

File: srcpkgs/arb/template
Comment (tornaria) [reply srcpkgs/arb/template]: - ntl has a custom configure perl script which runs binaries built for target.
- flintlib and arb should cross-compile fine once ntl is packaged

File: srcpkgs/eclib/template
Comment (tornaria) [reply srcpkgs/eclib/template]: `autoreconf -fi` is ok so I changed to that.

File: srcpkgs/flintlib/update
Comment (tornaria) [reply srcpkgs/flintlib/update:2]: made a comment in the `update` file itself, hope that's ok

File: srcpkgs/gf2x/template
Comment (tornaria) [reply srcpkgs/gf2x/template]: done for all 5 templates

File: srcpkgs/ntl/update
Comment (tornaria) [reply srcpkgs/ntl/update]: I changed site to be explicit and not use `${homepage}`, that's what seems to be done in most templates. I didn't see any notice in the template about syncing with `update`, but I think if the homepage changes for some reason and update-check can't find site anymore it will complain about "no version found" or something like that.

**Comment by tornaria** [comment]
Sorry for the double force-push. I realized a minute too late that I was missing a requested change.

I believe I have addressed all the suggestions.

**Comment by tornaria** [comment]
The "fix" for the `_GNU_SOURCE` issue was bad, which caused the musl build to fail check.

Based on this and the comments in https://github.com/wbhart/flint2/pull/988, I removed all patches and reverted back to using `CFLAGS=-D_GNU_SOURCE` in the template, with the care that for `arb` it has to be conditional to avoid a conflict.

Now all tests should pass... :crossed_fingers:

File: srcpkgs/gf2x/template
Comment (ericonr) [inline srcpkgs/gf2x/template:32]: `vmove "usr/lib/*.a"`

File: srcpkgs/ntl/template
Comment (ericonr) [inline srcpkgs/ntl/template:33]: same

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:25]: Same

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:27]: Same

File: srcpkgs/flintlib/template
Comment (ericonr) [inline srcpkgs/flintlib/template:20]: Make this  a single line

File: srcpkgs/arb/template
Comment (ericonr) [inline srcpkgs/arb/template:22]: Same here

File: srcpkgs/eclib/template
Comment (ericonr) [inline srcpkgs/eclib/template:29]: Same here

**Comment by tornaria** [comment]
All done.

**Review by ericonr — APPROVED** [review]
LGTM, going to leave open for a bit so anyone can leave thoughts before merging.

**Comment by tornaria** [comment]
@dkwo: do you want to test this? I've successfully built sagemath-9.4.rc1 with new pari + new giac + this PR.

Note that you have to rebase this PR before compiling so it gets the pari update (otherwise it will complain about a missing soname and refuse to pkg).

**Comment by dkwo** [comment]
@tornaria Thanks, I noticed that.
Didn't you need `configure_args+=" --with-readline --with-gmp"` in pari to use with sage?
I will rebuild sage as well soon, but I think you can go ahead with these PR's in the meantime.

**Comment by tornaria** [comment]
> @tornaria Thanks, I noticed that.
> Didn't you need `configure_args+=" --with-readline --with-gmp"` in pari to use with sage?

Not really, they are default on when `readline-devel` and `gmp-devel` are installed.

If you start our new gp (already in the void repo) you can see in the banner:
```
                                                                 GP/PARI CALCULATOR Version 2.13.2 (released)
                                                         amd64 running linux (x86-64/GMP-6.2.1 kernel) 64-bit version
                                                           compiled: Aug 15 2021, gcc version 10.2.1 20201203 (GCC)
                                                                          threading engine: pthread
                                                                (readline v8.1 enabled, extended help enabled)
```
Here gmp, readline are enabled. Also pthreads, which is required in sage starting in 9.4.

**Comment by dkwo** [comment]
I see, you are right.

**Comment by dkwo** [comment]
@tornaria Btw, I've also just built `sage-9.4.rc2` with this, the other four math package and the giac PR's :)

**Comment by ericonr** [comment]
Ping me for the next math package PR we should review ;)

**Pull request merged**
